includes/process-purchase.php

```diff
@@ -746,81 +753,99 @@ function edd_purchase_form_validate_guest_user() {
 }
 
 /**
- * Register And Login New User
+ * Register And Login New User.
  *
- * @param array   $user_data
+ * @since 1.0.8.1
  *
- * @access  private
- * @since  1.0.8.1
- * @return  integer
+ * @param array $user_data The data provided by the checkout page's registration form.
+ * @return integer
  */
 function edd_register_and_login_new_user( $user_data = array() ) {
-	// Verify the array
-	if ( empty( $user_data ) )
+
+	// Verify the array.
+	if ( empty( $user_data ) ) {
 		return -1;
+	}
 
-	if ( edd_get_errors() )
+	if ( edd_get_errors() ) {
 		return -1;
+	}
+
+	$user_args = apply_filters(
+		'edd_insert_user_args',
+		array(
+			'user_login'      => isset( $user_data['user_login'] ) ? $user_data['user_login'] : '',
+			'user_pass'       => isset( $user_data['user_pass'] ) ? $user_data['user_pass'] : '',
+			'user_email'      => isset( $user_data['user_email'] ) ? $user_data['user_email'] : '',
+			'first_name'      => isset( $user_data['user_first'] ) ? $user_data['user_first'] : '',
+			'last_name'       => isset( $user_data['user_last'] ) ? $user_data['user_last'] : '',
+			'user_registered' => date( 'Y-m-d H:i:s' ),
+			'role'            => get_option( 'default_role' ),
+		),
+		$user_data
+	);
 
-	$user_args = apply_filters( 'edd_insert_user_args', array(
-		'user_login'      => isset( $user_data['user_login'] ) ? $user_data['user_login'] : '',
-		'user_pass'       => isset( $user_data['user_pass'] )  ? $user_data['user_pass']  : '',
-		'user_email'      => isset( $user_data['user_email'] ) ? $user_data['user_email'] : '',
-		'first_name'      => isset( $user_data['user_first'] ) ? $user_data['user_first'] : '',
-		'last_name'       => isset( $user_data['user_last'] )  ? $user_data['user_last']  : '',
-		'user_registered' => date( 'Y-m-d H:i:s' ),
-		'role'            => get_option( 'default_role' )
-	), $user_data );
-
-	// Insert new user
+	// Insert new user.
 	$user_id = wp_insert_user( $user_args );
 
-	// Validate inserted user
-	if ( is_wp_error( $user_id ) )
+	// Validate inserted user.
+	if ( is_wp_error( $user_id ) ) {
 		return -1;
+	}
 
-	// Allow themes and plugins to filter the user data
+	// Allow themes and plugins to filter the user data.
 	$user_data = apply_filters( 'edd_insert_user_data', $user_data, $user_args );
 
-	// Allow themes and plugins to hook
+	// Allow themes and plugins to hook.
 	do_action( 'edd_insert_user', $user_id, $user_data );
 
-	// Login new user
+	// Login new user.
 	edd_log_user_in( $user_id, $user_data['user_login'], $user_data['user_pass'] );
 
-	// Return user id
+	// Since wp_signon doesn't set the current user, we need to do this.
+	wp_set_current_user( $user_id );
```

Comment:
Should we set line 803 to be a variable assignment, since `edd_log_user_in` will return either a `WP_User` or `WP_Error` object (what `wp_signon` returns), and then only set the current user if that is a `WP_User` object? It may not be necessary, though, as it looks like `wp_set_current_user` will perform the check itself, if the `$user_id` here is not a valid user.

Reply:
Related to `edd_log_user_in()` it might make more sense to always call that in that function? `wp_signon()` calls `wp_set_auth_cookie()` so always calling `wp_set_current_user()` inside of `edd_log_user_in()` would be closer to the previous behavior: https://github.com/easydigitaldownloads/easy-digital-downloads/commit/b7efaf07d5ce2b0ba23ca07fb5bec22bc878f173#diff-bf871927add1d88bdfed2583a5816e54L121